clm/trainer_utils_gpu.py

```diff
@@ -0,0 +1,302 @@
+
+    def train(self):
+        if (not isinstance(self.train_dataloader.batch_sampler, MegatronPretrainingRandomBatchSampler)) and (
+            self.cfg.max_epochs is not None and self.cfg.max_epochs > 1
+        ):
+            # if you use MegatronPretrainingBatchSampler as the batch_sampler passed to your train dataloader (in builders.py)
+            # then each epoch will repeat all your samples in the same order as the previous epoch, there is no shuffling
+            # to fix this, you should use MegatronPretrainingRandomBatchSampler instead, which alleviates this issue and allows
+            # random shuffling for each epoch.
+            raise ValueError(
+                "max_epochs > 1 is not supported unless using `MegatronPretrainingRandomBatchSampler` as the batch_sampler for your train dataloader"
+            )
+
+        epoch_iter = range(self.epoch, self.cfg.max_epochs)
+        if len(epoch_iter) <= 0:
+            # epoch done
+            return
+
+        self.run_timer.start_time()
+
+        if self.run_init_validation:
+            logging.info("Running validation in the very beginning.")
+            val_loss, val_metrics = self.run_validation()
+            # validation is done on the UPDATED weights
+            # so we use the incremented self.step
+            self.logger.log_metrics(val_metrics, step=self.step, prefix="val/")
+            logging.info("Initial validation metrics logged.")
+
+        for _ in epoch_iter:
+            num_steps_in_epoch = min(
+                self.max_steps - self.step, self.num_steps_per_epoch - self.step % self.num_steps_per_epoch
+            )
+            loop_iter = range(num_steps_in_epoch)
+
+            if not loop_iter:
+                return  # training ended
+
+            global_pbar = tqdm(
+                self.train_dataloader, initial=self.step, total=self.max_steps, leave=True, desc="Training steps"
+            )
+
+            for _, batch in zip(loop_iter, global_pbar):
+
+                self.timer.start("train_step_time")
+                loss, metrics = self.train_single_step(batch)
+                self.timer.stop("train_step_time")
+                train_step_time = self.timer.get("train_step_time")
+
+                # TODO(geshen): maybe use the dataloader instead
+                self.consumed_samples += self.model.cfg.global_batch_size
+                metrics["consumed_samples"] = self.consumed_samples
+                metrics["step_time"] = train_step_time
+                metrics["epoch"] = self.epoch + 1
+                self.logger.log_metrics(
+                    metrics, step=self.step, prefix="train/",
+                )
+                metrics = {f"train_{k}": v for k, v in metrics.items()}
+
+                self.step += 1
+
+                run_time_exceeded = self.run_timer.is_finished()
+                run_val, save_model, is_train_end = check_progress(
```

Comment:
LGTM.

Looks like `check_progress` is used for model saving only. Let's keep it though we probably don't need to save checkpoint in the middle.